### ISSUE TITLE
envoy command deadlocks when pipe buffers are too small

### ISSUE THREAD

**Issue by eoinburke67** [hashicorp/consul#10246]
**[edit]** This was originally reported on consul-k8s but transferred as it's really a core consul issue.

The tl;dr is that if the pipe buffer is reduced to one page, or if there is an abnormally large bootstrap file the command can block. There are several linux tunables that can affect pipe buffer sizes, linux capabilites can also have an effect, but even by default a user running lots of programs using pipes may have each one reduced to a single page of buffer, so the presence of other unrelated processes on a host may also cause this. 

If you can work around the cause and restore a large-enough pipe buffer then the command will work as-is, this issue remains open until we find an alternative mechanism that will work regardless of pipe buffer size.

-----

When filing a bug, please include the following headings if possible. Any example text in this template can be deleted.

### Overview of the Issue

After increasing the resources allocated to the lifecycle sidecar, the restart of the deployment rendered the mesh-gateway unable to start. 

### Reproduction Steps

unable to reproduce out side of the environment it is happening

1. The deployment has been running some months with `values.yml`:
```
global:
  name: consul
  # image: consul:1.8.0-beta1
  image: consul:1.8.4
  imageK8S: consul-k8s:0.19.0
  imageEnvoy: envoy-alpine:v1.14.4
  datacenter: xdcr-dc1
  enablePodSecurityPolicies: true

  # TLS configures whether Consul components use TLS.
  tls:
    # TLS must be enabled for federation in Kubernetes.
    enabled: true

  lifecycleSidecarContainer:
    resources:
      limits:
        memory: "100Mi"
        cpu: "50m"  
  
  federation:
    enabled: true
    # This will cause a Kubernetes secret to be created that
    # can be imported by secondary datacenters to configure them
    # for federation.
    createFederationSecret: true

  # # Gossip encryption secures the protocol Consul uses to quickly
  # # discover new nodes and detect failure.
  gossipEncryption:
    secretName: consul-gossip-encryption-key
    secretKey: key

connectInject:
  # Consul Connect service mesh must be enabled for federation.
  enabled: true

meshGateway:
  # Mesh gateways are gateways between datacenters. They must be enabled
  # for federation in Kubernetes since the communication between datacenters
  # goes through the mesh gateways.
  enabled: true
  replicas: 1
  service:
    port: 8081
    annotations: |
      service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
      service.beta.kubernetes.io/aws-load-balancer-internal: "true"

client:
  tolerations: |
    - key: "openet.com/voltdb"
      operator: "Equal"
      value: "true"
      effect: "NoSchedule"
  resources:
    requests:
      memory: "100Mi"
      cpu: "100m"
    limits:
      memory: "500Mi"
      cpu: "500m"

server:
  replicas: 1
  bootstrapExpect: 1
  storageClass: gp2
  nodeSelector: "transparent_hugepage: disabled"
  tolerations: |
    - key: "openet.com/voltdb"
      operator: "Equal"
      value: "true"
      effect: "NoSchedule"
  resources:
    requests:
      memory: "100Mi"
      cpu: "100m"
    limits:
      memory: "900Mi"
      cpu: "900m"

syncCatalog:
    enabled: false

ui:
  enabled: false
  service:
    type: 'LoadBalancer'
    annotations: |
      service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
      service.beta.kubernetes.io/aws-load-balancer-internal: "true"
```

Chart:
```
apiVersion: v1
appVersion: 1.8.4
description: Official HashiCorp Consul Chart
home: https://www.consul.io
icon: https://raw.githubusercontent.com/hashicorp/consul-helm/master/assets/icon.png
name: consul
sources:
- https://github.com/hashicorp/consul
- https://github.com/hashicorp/consul-helm
- https://github.com/hashicorp/consul-k8s
version: 0.25.0
```
1. View error
Note I just restarted the pod, but it is crashing for days 
```
[eburke@bastion ~]$ k get pods
NAME                                                          READY   STATUS    RESTARTS   AGE
consul-4gxfv                                                  1/1     Running   0          28d
consul-b8ccf                                                  1/1     Running   0          28d
consul-chh9j                                                  1/1     Running   0          28d
consul-connect-injector-webhook-deployment-55c885d56f-t9x5m   1/1     Running   0          2d12h
consul-hr27d                                                  1/1     Running   0          2d12h
consul-mesh-gateway-5d8fd96bc8-jm682                          1/2     Running   5          11m
consul-r9x9v                                                  1/1     Running   0          28d
consul-server-0                                               1/1     Running   0          28d
```

### Mesh Gateway logs

```
[eburke@bastion ~]$ k logs consul-mesh-gateway-5d8fd96bc8-jm682 copy-consul-bin
```
```
[eburke@bastion ~]$ k logs consul-mesh-gateway-5d8fd96bc8-jm682 service-init
Address "ae83295f223114efc9002314c56521e1-679905e2cc0fe953.elb.eu-central-1.amazonaws.com" written to /tmp/address.txt successfully
Registered service: mesh-gateway
```
```
[eburke@bastion ~]$ k logs consul-mesh-gateway-5d8fd96bc8-jm682 lifecycle-sidecar
2021-04-28T21:17:14.467Z [INFO]  Command configuration: service-config=/consul/service/service.hcl consul-binary=/consul-bin/consul sync-period=10s log-level=info
2021-04-28T21:17:22.466Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:17:39.766Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:17:56.668Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:18:12.768Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:18:29.367Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:18:46.366Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:19:03.566Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:19:21.966Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:19:39.167Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:19:55.867Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:20:11.867Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:20:28.766Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:20:45.467Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:21:02.866Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:21:20.266Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:21:36.266Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:21:53.067Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:22:09.367Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:22:25.868Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:22:42.467Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:22:59.466Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:23:16.867Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:23:33.566Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:23:50.267Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:24:06.468Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:24:23.666Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:24:40.667Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:24:58.366Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:25:14.966Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:25:31.866Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:25:49.066Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:26:05.466Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:26:22.166Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:26:37.867Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:26:54.767Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:27:12.167Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:27:29.466Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:27:46.266Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:28:04.067Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:28:20.467Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:28:36.567Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:28:52.566Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:29:10.066Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
2021-04-28T21:29:26.266Z [INFO]  successfully synced service: output="Registered service: mesh-gateway"
```
```
[eburke@bastion ~]$ k logs consul-mesh-gateway-5d8fd96bc8-jm682 mesh-gateway
==> No mesh-gateway services registered with this agent
```

I have metric beat, part of the elastic stack for pod metrics and the pods usage is always below 50 m/Mi

### Expected behavior

Mesh gateway starts

### Environment details
AWS
EKS
```
[eburke@bastion ~]$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.4", GitCommit:"c96aede7b5205121079932896c4ad89bb93260af", GitTreeState:"clean", BuildDate:"2020-06-17T11:41:22Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.17", GitCommit:"68b4e26caf6ede7af577db4af62fb405b4dd47e6", GitTreeState:"clean", BuildDate:"2021-03-18T00:54:02Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
### Additional Context

This had been working for some months without issue

### What I have tried so far
1. Scaling back the deployment to 0 and creating the pod from fresh
2. restarting the mesh-gateway, consul client supporting it and the consul server
3. force deregistering the mesh-gateway and restarting the pod
4. increasing the resources of the mesh-gateway to 5GB/6CPU
5. increasing failure thresholds to 10 to allow for a longer startup

### pod description
```
[eburke@bastion ~]$ k describe pod consul-mesh-gateway-5d8fd96bc8-jm682
Name:         consul-mesh-gateway-5d8fd96bc8-jm682
Namespace:    infrastructure-test-fra
Priority:     0
Node:         ip-10-1-24-67.eu-central-1.compute.internal/10.1.24.67
Start Time:   Wed, 28 Apr 2021 21:17:04 +0000
Labels:       app=consul
              chart=consul-helm
              component=mesh-gateway
              pod-template-hash=5d8fd96bc8
              release=consul
Annotations:  consul.hashicorp.com/connect-inject: false
              kubernetes.io/psp: consul-mesh-gateway
Status:       Running
IP:           10.170.4.22
IPs:
  IP:           10.170.4.22
Controlled By:  ReplicaSet/consul-mesh-gateway-5d8fd96bc8
Init Containers:
  copy-consul-bin:
    Container ID:  docker://fbeddeecdf8c96be4ee78522bc7a837e8c606e0dbc0960a7eaa0d11a6c9fa739
    Image:         .dkr.ecr.eu-central-1.amazonaws.com/consul:1.8.4
    Image ID:      docker-pullable://.dkr.ecr.eu-central-1.amazonaws.com/consul@sha256:efa2713f11972127ecca593aaf68bfb9609412ca9d3ee1892ee932a29c3e70af
    Port:          <none>
    Host Port:     <none>
    Command:
      cp
      /bin/consul
      /consul-bin/consul
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 28 Apr 2021 21:17:05 +0000
      Finished:     Wed, 28 Apr 2021 21:17:06 +0000
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     50m
      memory:  150Mi
    Requests:
      cpu:        50m
      memory:     25Mi
    Environment:  <none>
    Mounts:
      /consul-bin from consul-bin (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from consul-mesh-gateway-token-nxnds (ro)
  service-init:
    Container ID:  docker://8975fe29fb298368e8bcbc21ddafbf0e0862b10c143189f9e21a4371f5ade3cb
    Image:         .dkr.ecr.eu-central-1.amazonaws.com/consul-k8s:0.19.0
    Image ID:      docker-pullable://.dkr.ecr.eu-central-1.amazonaws.com/consul-k8s@sha256:87cf443ddca2553a05fcf7156e09e99eebf59a028f555e1a0fddef7874cafae1
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      -ec
      consul-k8s service-address \
        -k8s-namespace=infrastructure-test-fra \
        -name=consul-mesh-gateway \
        -output-file=/tmp/address.txt
      WAN_ADDR="$(cat /tmp/address.txt)"
      WAN_PORT="8081"

      cat > /consul/service/service.hcl << EOF
      service {
        kind = "mesh-gateway"
        name = "mesh-gateway"
        meta {
          consul-wan-federation = "1"
        }
        port = 8443
        address = "${POD_IP}"
        tagged_addresses {
          lan {
            address = "${POD_IP}"
            port = 8443
          }
          wan {
            address = "${WAN_ADDR}"
            port = ${WAN_PORT}
          }
        }
        checks = [
          {
            name = "Mesh Gateway Listening"
            interval = "10s"
            tcp = "${POD_IP}:8443"
            deregister_critical_service_after = "6h"
          }
        ]
      }
      EOF

      /consul-bin/consul services register \
        /consul/service/service.hcl

    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 28 Apr 2021 21:17:07 +0000
      Finished:     Wed, 28 Apr 2021 21:17:11 +0000
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     50m
      memory:  25Mi
    Requests:
      cpu:     50m
      memory:  25Mi
    Environment:
      HOST_IP:            (v1:status.hostIP)
      POD_IP:             (v1:status.podIP)
      CONSUL_HTTP_ADDR:  https://$(HOST_IP):8501
      CONSUL_CACERT:     /consul/tls/ca/tls.crt
    Mounts:
      /consul-bin from consul-bin (rw)
      /consul/service from consul-service (rw)
      /consul/tls/ca from consul-ca-cert (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from consul-mesh-gateway-token-nxnds (ro)
Containers:
  mesh-gateway:
    Container ID:  docker://7f34cf628843a4d0c68f310a4aecf3447a8e965589e8dbe7b8eeb031cd097ce8
    Image:         .dkr.ecr.eu-central-1.amazonaws.com/envoy-alpine:v1.14.4
    Image ID:      docker-pullable://.dkr.ecr.eu-central-1.amazonaws.com/envoy-alpine@sha256:9442ccf7b335045ab5ef596e26bec978b8ef46f8bce1dfeb19382c3f40208eb5
    Port:          8443/TCP
    Host Port:     0/TCP
    Command:
      /consul-bin/consul
      connect
      envoy
      -mesh-gateway
      --
      -l
      debug
      -log-level=debug
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    137
      Started:      Wed, 28 Apr 2021 21:31:48 +0000
      Finished:     Wed, 28 Apr 2021 21:34:04 +0000
    Ready:          False
    Restart Count:  7
    Limits:
      cpu:     7
      memory:  5000Mi
    Requests:
      cpu:      6
      memory:   5000Mi
    Liveness:   tcp-socket :8443 delay=30s timeout=5s period=10s #success=1 #failure=10
    Readiness:  tcp-socket :8443 delay=10s timeout=5s period=10s #success=1 #failure=3
    Environment:
      HOST_IP:            (v1:status.hostIP)
      POD_IP:             (v1:status.podIP)
      CONSUL_HTTP_ADDR:  https://$(HOST_IP):8501
      CONSUL_GRPC_ADDR:  https://$(HOST_IP):8502
      CONSUL_CACERT:     /consul/tls/ca/tls.crt
    Mounts:
      /consul-bin from consul-bin (rw)
      /consul/tls/ca from consul-ca-cert (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from consul-mesh-gateway-token-nxnds (ro)
  lifecycle-sidecar:
    Container ID:  docker://e5f8986140278961a2fd6a6905ee6ddf335258736e59e43e887cf2cf8de5932a
    Image:         .dkr.ecr.eu-central-1.amazonaws.com/consul-k8s:0.19.0
    Image ID:      docker-pullable://.dkr.ecr.eu-central-1.amazonaws.com/consul-k8s@sha256:87cf443ddca2553a05fcf7156e09e99eebf59a028f555e1a0fddef7874cafae1
    Port:          <none>
    Host Port:     <none>
    Command:
      consul-k8s
      lifecycle-sidecar
      -service-config=/consul/service/service.hcl
      -consul-binary=/consul-bin/consul
    State:          Running
      Started:      Wed, 28 Apr 2021 21:17:12 +0000
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     20m
      memory:  50Mi
    Requests:
      cpu:     20m
      memory:  25Mi
    Environment:
      HOST_IP:            (v1:status.hostIP)
      POD_IP:             (v1:status.podIP)
      CONSUL_HTTP_ADDR:  https://$(HOST_IP):8501
      CONSUL_CACERT:     /consul/tls/ca/tls.crt
    Mounts:
      /consul-bin from consul-bin (rw)
      /consul/service from consul-service (ro)
      /consul/tls/ca from consul-ca-cert (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from consul-mesh-gateway-token-nxnds (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  consul-bin:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  consul-service:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  consul-ca-cert:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  consul-ca-cert
    Optional:    false
  consul-mesh-gateway-token-nxnds:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  consul-mesh-gateway-token-nxnds
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason     Age                   From                                                  Message
  ----     ------     ----                  ----                                                  -------
  Normal   Scheduled  <unknown>             default-scheduler                                     Successfully assigned infrastructure-test-fra/consul-mesh-gateway-5d8fd96bc8-jm682 to ip-10-1-24-67.eu-central-1.compute.internal
  Normal   Pulling    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Pulling image ".dkr.ecr.eu-central-1.amazonaws.com/consul:1.8.4"
  Normal   Pulled     18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Successfully pulled image ".dkr.ecr.eu-central-1.amazonaws.com/consul:1.8.4"
  Normal   Created    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Created container copy-consul-bin
  Normal   Started    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Started container copy-consul-bin
  Normal   Pulling    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Pulling image ".dkr.ecr.eu-central-1.amazonaws.com/consul-k8s:0.19.0"
  Normal   Pulled     18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Successfully pulled image ".dkr.ecr.eu-central-1.amazonaws.com/consul-k8s:0.19.0"
  Normal   Created    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Created container service-init
  Normal   Started    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Started container service-init
  Normal   Pulling    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Pulling image ".dkr.ecr.eu-central-1.amazonaws.com/envoy-alpine:v1.14.4"
  Normal   Pulled     18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Successfully pulled image ".dkr.ecr.eu-central-1.amazonaws.com/envoy-alpine:v1.14.4"
  Normal   Created    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Created container mesh-gateway
  Normal   Started    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Started container mesh-gateway
  Normal   Pulling    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Pulling image ".dkr.ecr.eu-central-1.amazonaws.com/consul-k8s:0.19.0"
  Normal   Pulled     18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Successfully pulled image ".dkr.ecr.eu-central-1.amazonaws.com/consul-k8s:0.19.0"
  Normal   Created    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Created container lifecycle-sidecar
  Normal   Started    18m                   kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Started container lifecycle-sidecar
  Warning  Unhealthy  17m (x4 over 17m)     kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Liveness probe failed: dial tcp 10.170.4.22:8443: connect: connection refused
  Warning  Unhealthy  3m25s (x69 over 18m)  kubelet, ip-10-1-24-67.eu-central-1.compute.internal  Readiness probe failed: dial tcp 10.170.4.22:8443: connect: connection refused
```

**Comment by fubinl** [comment]
Hi @banks,
Is there a timeline on when will this issue be resolved? And will the potential solution back ported to the earlier releases e.g. 1.8.x?
Thanks.

**Comment by dnephin** [comment]
#10324 should fix this issue
